dxr/vcs.py

```diff
@@ -327,8 +329,123 @@ def get_contents(cls, path, revision, stderr=None):
+        self.files = self.__get_files_with_revision()
+        # Determine repository revision once.
+        self.revision = self.__get_revision()
+        # Determine upstream URL once.
+        self.upstream = self.__get_upstream_url()
+
+    # Interface class methods.
+
+    @classmethod
+    def claim_vcs_source(cls, path, dirs, tree):
+        # Requires ".svn" directory.
+        if '.svn' not in dirs:
+            return None
+
+        # Requires successful "svn info".
+        try:
+            cls.invoke_vcs(['info'], path)
+        except subprocess.CalledProcessError:
+            return None
+
+        # Exclude ".svn" from indexing.
+        dirs.remove('.svn')
+
+        return cls(path)
+
+    @classmethod
+    def get_contents(cls, path, revision, stderr=None):
+        directory, filename = split(path)
+        args = ['cat', '-r', str(revision), os.path.abspath(path)]
+        return cls.invoke_vcs(args, directory)
+
+    # Interface instance methods.
+
+    def is_tracked(self, path):
+        return path in self.files
+
+    def has_upstream(self):
+        return self.upstream != ""
+
+    def generate_log(self, path):
+        return None
+
+    def generate_diff(self, path):
+        return None
+
+    def generate_blame(self, path):
+        return None
+
+    def generate_raw(self, path):
+        return self.upstream + "/" + path
+
+    def display_rev(self, path):
+        return "r%s" % self.files[path]
+
+    # Private methods.
+
+    def __get_files_with_revision(self):
+        files = dict()
+
+        root_path = "%s/" % self.root
+
+        for root, dirnames, filenames in os.walk(self.root):
+            for filename in filenames:
+                abs_path = join(root, filename)
+
+                if not isfile(abs_path):
+                    # Ignore directories.
+                    continue
+
+                rel_path = abs_path.replace(root_path, '')
 
-every_vcs = [Mercurial, Git, Perforce]
+                if rel_path.startswith('.svn'):
+                    # Ignore ".svn" path.
+                    continue
+
+                revision = self.__get_last_changed_rev(abs_path)
+
+                if not revision:
+                    # Ignore untracked files.
+                    continue
+
+                files[rel_path] = revision
+
+        return files
+
+    def __get_revision(self):
+        return Subversion.svn_info(self.root, "Revision")
+
+    def __get_upstream_url(self):
+        return Subversion.svn_info(self.root, "URL")
+
+    def __get_last_changed_rev(self, path):
+        return Subversion.svn_info(self.root, "Last Changed Rev", [path])
+
+    @classmethod
+    def svn_info(cls, path, prop=None, args=[]):
+        try:
+            info = cls.invoke_vcs(['info'] + args, path)
+
+            if not prop:
+                # Return full info.
+                return info
+
+            # Find and return specific property.
+            regex = re.compile("(?<=%s: )(.+)" % prop)
```

Comment:
Just to be safe, it might make sense to re.escape(prop)

Reply:
Sorry for the delay, this is now fixed in 12eb447f.